Refactor unit tests to use libvmi for resource creation

aliamerj opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
Before this PR:
- Unit tests directly created and modified VM and VMI resources by manipulating their fields.
After this PR:
- Unit tests have been refactored to use the `libvmi` package for creating and modifying VM and VMI resources, promoting readability and standardization.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->
Partially address: https://github.com/kubevirt/kubevirt/issues/12059

### Why we need it and why it was done in this way
 - Using the libvmi package simplifies the unit tests and makes them more readable.
 - This approach ensures a standard way of creating VM and VMI resources across tests.
 
The following tradeoffs were made:
  - Direct manipulation of resources was replaced with utility functions, which may introduce minor overhead but enhances readability and maintainability.

The following alternatives were considered:
 - Continuing with direct manipulation of resources in tests, which was deemed less maintainable.
Links to places where the discussion took place: Issue discussion: [#12059](https://github.com/kubevirt/kubevirt/issues/12059)

### Special notes for your reviewer

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

aliamerj commented:
@alicefr @jean-edouard @fossedihelm
Could you please give me feedback on this PR when you have some free time?

Thank you!

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -92,78 +91,98 @@ var _ = Describe("ContainerDisk", func() {
 
 			It("by verifying error occurs if multiple host directory locations exist somehow", func() {
-				vmi := api.NewMinimalVMI("fake-vmi")
+				vmi := libvmi.New(
```
The original VMI did not have a containerdisk.

aliamerj replied:
yes you are right i just replaced it with 
```go
vmi := libvmi.New()
```
and still working
i got confused with name i thought i will get null error if i did not define the  containerdisk

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -49,9 +48,9 @@ var _ = Describe("ContainerDisk", func() {
 	VerifyDiskType := func(diskExtension string) {
-		vmi := api.NewMinimalVMI("fake-vmi")
+
+		vmi := libvmi.New(libvmi.WithContainerDisk("r0", "someimage:v1.2.3.4"))
```
nit: Use same indentation as below.

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -92,78 +91,98 @@ var _ = Describe("ContainerDisk", func() {
+				vmi := libvmi.New(
+					libvmi.WithContainerDisk("r0", "someimage:v1.2.3.4"),
+				)
+				By("Ensuring no disk is present")
+				// Don't create any disk file to simulate the condition
```
I'd drop this

aliamerj replied:
you mean drop only this right ??
```go
	By("Ensuring no disk is present")
				// Don't create any disk file to simulate the condition
```

0xFelix replied:
Yes, it's okay to leave this out.

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -92,78 +91,98 @@ var _ = Describe("ContainerDisk", func() {
+				By("Ensuring no disk is present")
+				// Don't create any disk file to simulate the condition
+
+				By("Trying to get the volume mount directory on guest and expecting an error")
+				volumeMountDir := GetVolumeMountDirOnGuest(vmi)
+				_, err := os.Stat(filepath.Join(volumeMountDir, "disk_0.img"))
+				Expect(err).To(HaveOccurred())
+				Expect(errors.Is(err, os.ErrNotExist)).To(BeTrue())
```
This was missing before, correct?

aliamerj replied:
correct 
It was just and i did not see the assertion
```go
It("by verifying error when no disk is present", func() {
vmi := api.NewMinimalVMI("fake-vmi")
appendContainerDisk(vmi, "r0")
})
```
i just add extra assertion , just to make sure everything is working fine and make sense but if you see no need i can remove it
```go
It("by verifying error when no disk is present", func() {
vmi := libvmi.New()

By("Trying to get the volume mount directory on guest and expecting an error")
volumeMountDir := GetVolumeMountDirOnGuest(vmi)
_, err := os.Stat(filepath.Join(volumeMountDir, "disk_0.img"))
Expect(err).To(HaveOccurred())
Expect(errors.Is(err, os.ErrNotExist)).To(BeTrue())
})

what do you think ??

alicefr replied:
I think here would be more indicate to use `MatchError` -> `Expect(err).To(MatchError(os.ErrNotExist))`

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -92,78 +91,98 @@ var _ = Describe("ContainerDisk", func() {
 
 			It("by verifying launcher directory locations", func() {
-				vmi := api.NewMinimalVMI("fake-vmi")
+				vmi := libvmi.New(
+					libvmi.WithContainerDisk("disk1", "someimage:v1.2.3.4"),
+				)
```
Same, original VMI did not have a containerdisk.

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -92,78 +91,98 @@ var _ = Describe("ContainerDisk", func() {
 
 			It("by verifying host directory locations", func() {
-				vmi := api.NewMinimalVMI("fake-vmi")
+				vmi := libvmi.New(
+					libvmi.WithContainerDisk("r0", "someimage:v1.2.3.4"),
+				)
```
Same

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -182,8 +201,13 @@ var _ = Describe("ContainerDisk", func() {
-				appendContainerDisk(vmi, "r0")
+				vmi := libvmi.New(
+					libvmi.WithContainerDisk("r0", "someimage:v1.2.3.4"),
+					libvmi.WithResourceMemory("64M"),
+					libvmi.WithResourceCPU("1"),
```
After adding this, are the lines below still needed?

aliamerj replied:
which lines  ?

aliamerj replied:
yes , i already removed these lines ( if you mean these line )
```go
By("Setting the VMI's resource requirements")
vmi.Spec.Domain.Resources = v1.ResourceRequirements{
Requests: k8sv1.ResourceList{
k8sv1.ResourceCPU:    resource.MustParse("1"),
k8sv1.ResourceMemory: resource.MustParse("64M"),
},
Limits: k8sv1.ResourceList{
k8sv1.ResourceCPU:    resource.MustParse("1"),
k8sv1.ResourceMemory: resource.MustParse("64M"),
},
}
```
and I got this error 
```bash
meline >>
  {"component":"","level":"info","msg":"Updating cluster config from KubeVirt to resource version 'm66xmkq9mm'","pos":"configuration.go:365","timestamp":"2024-06-14T11:50:16.416341Z"}
  [FAILED] in [It] - /home/ali/Projects/kubevirt/pkg/container-disk/container-disk_test.go:201 @ 06/14/24 14:50:16.416
  << Timeline

  [FAILED] Expected
      <v1.ResourceList | len:3>: {
          "ephemeral-storage": {
              i: {value: 50, scale: 6},
              d: {Dec: nil},
              s: "50M",
              Format: "DecimalSI",
          },
          "cpu": {
              i: {value: 1, scale: -3},
              d: {Dec: nil},
              s: "1m",
              Format: "DecimalSI",
          },
          "memory": {
              i: {value: 1, scale: 6},
              d: {Dec: nil},
              s: "1M",
              Format: "DecimalSI",
          },
      }
  to contain elements
      <[]resource.Quantity | len:3, cap:3>: [
          {
              i: {value: 10, scale: -3},
              d: {Dec: nil},
              s: "10m",
              Format: "DecimalSI",
          },
          {
              i: {value: 40, scale: 6},
              d: {Dec: nil},
              s: "40M",
              Format: "DecimalSI",
          },
          {
              i: {value: 50, scale: 6},
              d: {Dec: nil},
              s: "50M",
              Format: "DecimalSI",
          },
      ]
  the missing elements were
      <[]resource.Quantity | len:2, cap:2>: [
          {
              i: {value: 10, scale: -3},
              d: {Dec: nil},
              s: "10m",
              Format: "DecimalSI",
          },
          {
              i: {value: 40, scale: 6},
              d: {Dec: nil},
              s: "40M",
              Format: "DecimalSI",
          },
      ]

```

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -272,36 +300,50 @@ var _ = Describe("ContainerDisk", func() {
+				)
+
+				// Explicitly set ImagePullPolicy for each container disk
+				vmi.Spec.Volumes[0].ContainerDisk.ImagePullPolicy = k8sv1.PullAlways
```
What is the value that `libvmi.WithContainerDisk` is using by default? I see `appendContainerDisk` used `PullAlways`.

aliamerj replied:
i needed to set ImagePullPolicy to something because without it i am getting the nil error
```go
vmi.Spec.Volumes[0].ContainerDisk.ImagePullPolicy = k8sv1.PullAlways
```

alicefr replied:
I would add a new field in the method wit the pull policy, the goal here is to avoid to directly access the volumes

aliamerj replied:
so do i need to create new option ? 
like 
```go
libvmi.WithPullPolicy
```

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -272,36 +300,50 @@ var _ = Describe("ContainerDisk", func() {
 				containers := GenerateContainers(vmi, clusterConfig, nil, "libvirt-runtime", "bin-volume")
-				Expect(err).ToNot(HaveOccurred())
 
-				Expect(containers).To(HaveLen(2))
```
Do not drop this assertion.

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -346,43 +392,59 @@ var _ = Describe("ContainerDisk", func() {
+				)
+				vmi.Spec.Domain.Firmware = &v1.Firmware{
+					KernelBoot: &v1.KernelBoot{
+						Container: &v1.KernelBootContainer{Image: "someimage:v1.2.3.4"},
```
Is there a `libvmi` `Option` for this?

aliamerj replied:
if you mean for Firmware 
I tried to find the option but I couldn't
The only way was to set it explicitly

alicefr replied:
We should create the option in libvmi, you can look at some examples in `pkg/libvmi`

aliamerj replied:
okay i am really exciting to do that
it maybe better to complete this task first and than i will create options for `Firmware` and `WithPullPolicy` and i will back here and update these fields 
so what do you think ?

alicefr replied:
You could first create the 2 options in 2 separate commits and add them on top of this PR and then modify the tests accordingly.

aliamerj replied:
I feel i will mess up with commits and branches , Is it okay if did it one by one
fixing the task in hand and then implement the options and just get back and replace it will be much easy 
and i can make sure my implementation is correct

alicefr replied:
I think it is a good opportunity to learn git. I would suggest you to take some tutorials, then it becomes a common work flow.

0xFelix commented on pkg/host-disk/host-disk_test.go:
```diff
@@ -106,17 +80,23 @@ var _ = Describe("HostDisk", func() {
 	})
+	createHostDiskPath := func(volumeName string) string {
+		err := os.Mkdir(path.Join(tempDir, volumeName), 0755)
+		if !os.IsExist(err) {
+			Expect(err).NotTo(HaveOccurred())
+		}
```
I would like the following better:

```
if err != nil {
    Expect(err).ToMatchError(os.IsExist, "IsExist")
}
``

0xFelix replied:
Also, I would keep the old name (or change to `createHostDisk`) and add the capacity to this helper.

aliamerj replied:
what about 
```go
createHostDisk := func(volumeName string) string {
if err := os.Mkdir(path.Join(tempDir, volumeName), 0755); err != nil {
Expect(err).To(MatchError(os.ErrExist))	
}
return path.Join(tempDir, volumeName, "disk.img")
}
```

fossedihelm reviewed (commented):
Thank you for your PR. 
Few comments below.
Please rebase onto `main` instead of merge `main` in your branch.
Lastly, please update the PR description not using the word `Fixes`, otherwise that gh issue will be closed by this. Suggestion:
```
Partially address: https://github.com/kubevirt/kubevirt/issues/12059
```

alicefr commented:
@aliamerj please squash the merge commit, we need to have a clean git history. Additionally, could you please rewrite your commit message with a short title and then a brief explanation of the change in the commit body?

aliamerj commented:
sorry 
i will do it right now

alicefr commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -346,43 +392,59 @@ var _ = Describe("ContainerDisk", func() {
 				Expect(newContainers[0].Image).To(Equal("someimage@sha256:0"))
 				Expect(newContainers[1].Image).To(Equal("someimage@sha256:1"))
 			})
+
```
nit: extra empty line

alicefr commented on pkg/host-disk/host-disk_test.go:
```diff
@@ -148,19 +127,25 @@ var _ = Describe("HostDisk", func() {
+						libvmi.WithHostDisk("volume3", createHostDiskPath("volume3"), v1.HostDiskExistsOrCreate),
+					)
+					By("Setting the capacity for each HostDisk volume")
+					vmi.Spec.Volumes[0].HostDisk.Capacity = resource.MustParse("64Mi")
```
Same here we shouldn't directly access the volumes directly but instead create an Option in libvmi

0xFelix replied:
What about making the size an argument on `createHostDisk` for now?

aliamerj replied:
I’ve made some changes to enhance the reusability and flexibility of HostDisk related functions. Here’s a summary of the changes:
1. Created a new function WithHostDiskAndCapacity in `pkg/libvmi/storage.go`
This function allows specifying the capacity directly when creating a HostDisk:
```go
func WithHostDiskAndCapacity(diskName, path string, diskType v1.HostDiskType, capacity string) Option {
    return func(vmi *v1.VirtualMachineInstance) {
        addDisk(vmi, newDisk(diskName, v1.DiskBusVirtio))
        addVolume(vmi, newHostDisk(diskName, path, diskType, resource.MustParse(capacity)))
    }
}
```
2. Modified the `newHostDisk` function to accept an optional capacity argument
This makes newHostDisk more versatile, allowing it to handle both cases with or without specified capacity:
```go
func newHostDisk(name, path string, diskType v1.HostDiskType, capacity ...resource.Quantity) v1.Volume {
    hostDisk := v1.HostDisk{
        Path: path,
        Type: diskType,
    }

    // Set capacity if provided, otherwise default to defaultDiskSize
    if diskType == v1.HostDiskExistsOrCreate {
        if len(capacity) > 0 {
            hostDisk.Capacity = capacity[0]
        } else {
            hostDisk.Capacity = resource.MustParse(defaultDiskSize)
        }
    }

    return v1.Volume{
        Name: name,
        VolumeSource: v1.VolumeSource{
            HostDisk: &hostDisk,
        },
    }
}
```
and we can use it like this
```go
vmi := libvmi.New(
    libvmi.WithHostDiskAndCapacity("volume1", createHostDisk("volume1"), v1.HostDiskExistsOrCreate, "64Mi"),
    libvmi.WithHostDiskAndCapacity("volume2", createHostDisk("volume2"), v1.HostDiskExistsOrCreate, "128Mi"),
    libvmi.WithHostDiskAndCapacity("volume3", createHostDisk("volume3"), v1.HostDiskExistsOrCreate, "80Mi"),
)
```

0xFelix replied:
Looks like you missed to update L129-L138

aliamerj commented:
i believe i squashed the merge commit,

fossedihelm commented on pkg/host-disk/host-disk_test.go:
```diff
@@ -344,11 +341,10 @@ var _ = Describe("HostDisk", func() {
-			addHostDisk(vmi, "volume1", "UnknownType", "")
+			By("Creating a new VMI with a HostDisk volume and unknown type")
+			vmi := libvmi.New(
+				libvmi.WithHostDisk("volume1", createHostDisk("volume1"), ""),
```
```suggestion
				libvmi.WithHostDisk("volume1", createHostDisk("volume1"), "UnknownType"),
```

fossedihelm commented on pkg/host-disk/host-disk_test.go:
```diff
@@ -201,11 +190,13 @@ var _ = Describe("HostDisk", func() {
 
-					By("Adding HostDisk volumes")
-					addHostDisk(vmi, "volume1", v1.HostDiskExistsOrCreate, "64Mi")
+					By("Setting the capacity for HostDisk volume")
```
Again :)

fossedihelm reviewed (commented):
Generally, I think you lost some changes during the rebase, since most of my addressed comments are come back again.
I added some new comments below.
Thanks

aliamerj commented:
Yes , you are right , it gets overwritten i applied your suggestions 
Thank you for your review I really appreciate that

fossedihelm commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -182,8 +190,15 @@ var _ = Describe("ContainerDisk", func() {
+					libvmi.WithLimitCPU("1"),
+				)
+
+				By("Setting the VMI's resource requirements")
```
You can drop the explicit resource instructions below, since now u used libvmi.

0xFelix replied:
Looks like you reverted the libvmi options. Can you please use the appropriate options instead of modifying the VMI manually?

aliamerj replied:
make new option  `WithUID`  ??

fossedihelm reviewed (commented):
Can you squash the commits into a single one?
You can drop the Co-Authored part :)

fossedihelm commented on pkg/host-disk/host-disk_test.go:
```diff
@@ -148,19 +126,25 @@ var _ = Describe("HostDisk", func() {
 					Expect(err).NotTo(HaveOccurred())
 
-					// check if images exist and the size is adequate to requirements
+					// Ensure volumes are not nil and correctly initialized
+					Expect(vmi.Spec.Volumes).NotTo(BeNil())
```
Please drop this. If `vmi.Spec.Volumes` were nil, you got a nil pointer dereference error during:
```
vmi.Spec.Volumes[0].HostDisk.Capacity ...
```

fossedihelm commented on pkg/host-disk/host-disk_test.go:
```diff
@@ -286,10 +287,10 @@ var _ = Describe("HostDisk", func() {
+					By("Setting the capacity for each HostDisk volume")
+					vmi.Spec.Volumes[0].HostDisk.Capacity = resource.MustParse("64Mi")
+					vmi.Spec.Volumes[1].HostDisk.Capacity = resource.MustParse("64Mi")
+					vmi.Spec.Volumes[2].HostDisk.Capacity = resource.MustParse("1E")
```
?

aliamerj replied:
sorry, i fix it but it's been overwritten

fossedihelm reviewed (commented):
Still some small comments.
Thank you for your time

aliamerj commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -350,18 +383,23 @@ var _ = Describe("ContainerDisk", func() {
+
+				By("Creating a new VMI with container and non-container disks")
+				vmi := libvmi.New(
+					libvmi.WithContainerDisk("disk1", "someimage:v1.2.3.4"),
```
I used the `someimage:v1.2.3.4`  15 times in the file
I guess we could do that 
Do i need to replace all or just in this test ??

aliamerj commented:
@fossedihelm 
thank you very much for reviewing my code 
and i hope i did not waste your time

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -92,78 +91,88 @@ var _ = Describe("ContainerDisk", func() {
+				volumeMountDir := GetVolumeMountDirOnGuest(vmi)
+				_, err := os.Stat(filepath.Join(volumeMountDir, "disk_0.img"))
+				Expect(err).To(HaveOccurred())
+				Expect(err).To(MatchError(os.ErrNotExist))
```
Does this work? IIUC `MatchError` want's to see a string as second arg in uses like this?

aliamerj replied:
This method doesn't work anymore
the `ToMatchError` is undefined in `Expect(err).`
```go
Expect(err).ToMatchError(os.IsExist, "IsExist")]
```
we can achieve the same logic with
```go 
Expect(err).To(MatchError(os.ErrNotExist))
```

0xFelix replied:
https://onsi.github.io/gomega/#matcherrorexpected-interface

> If using a function in this way you are required to pass a FUNCTION_ERROR_DESCRIPTION argument to MatchError that describes the function. This description is used in the failure message. For example: Expect(err).To(MatchError(os.IsNotExist, "IsNotExist))

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -318,11 +348,15 @@ var _ = Describe("ContainerDisk", func() {
+						libvmi.WithContainerDisk("r1", someImage),
+						libvmi.WithContainerDisk("r2", someImage),
+					)
+					vmi.UID = "6789"
```
Where does this come from?

aliamerj replied:
I Create a new VMI with multiple container disks
instead of
```go
vmi = api.NewMinimalVMI("fake-vmi")
appendContainerDisk(vmi, "r0")
appendContainerDisk(vmi, "r1")
appendContainerDisk(vmi, "r2")
```
I used  libvmi
```go
vmi = libvmi.New(
libvmi.WithContainerDisk("r0", someImage),
libvmi.WithContainerDisk("r1", someImage),
libvmi.WithContainerDisk("r2", someImage),
)
```
I needed to provide to the WithContainerDisk image name 
```go
const someImage = "someimage:v1.2.3.4"
```
also I will remove the ``` vmi.UID = "6789" ```

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -370,32 +409,48 @@ var _ = Describe("ContainerDisk", func() {
+					libvmi.WithContainerDisk("disk1", someImage),
+					libvmi.WithDataVolume("disk3", "some-data-volume"),
+				)
+				vmi.Spec.Domain.Firmware = &v1.Firmware{
```
Potentially add a libvmi option for it.

aliamerj replied:
I'm working on it

0xFelix reviewed (commented):
/test all

aliamerj commented:
When I run make test, it changes the `BUILD.bazel` file. Is this behavior expected, or should I roll back the `BUILD.bazel` file to its original state?

fossedihelm commented:
> When I run make test, it changes the `BUILD.bazel` file. Is this behavior expected, or should I roll back the `BUILD.bazel` file to its original state?

This is an expected behavior. You have to run `make && make generate` and add all the changes to your commit.
Otherwise build will fail. Thank you

0xFelix commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -370,32 +404,44 @@ var _ = Describe("ContainerDisk", func() {
-				vmi.Spec.Domain.Firmware = &v1.Firmware{KernelBoot: &v1.KernelBoot{Container: &v1.KernelBootContainer{Image: "someimage:v1.2.3.4"}}}
+				By("Creating a new VMI with a container disk and a kernel boot image")
+				vmi := libvmi.New(
+					libvmi.WithFirmware(someImage),
```
It should be called `libvmi.WithKernelBootContainer`.

0xFelix commented on pkg/libvmi/storage.go:
```diff
@@ -219,15 +234,20 @@ func newVolume(name string) v1.Volume {
 
-func newContainerVolume(name, image string) v1.Volume {
-	return v1.Volume{
+func newContainerVolume(name, image string, imagePullPolicy ...k8sv1.PullPolicy) v1.Volume {
```
Can you pass `imagePullPolicy` as a string instead and check for empty string?

aliamerj replied:
okay

0xFelix commented on pkg/libvmi/storage.go:
```diff
@@ -40,6 +40,14 @@ func WithContainerDisk(diskName, imageName string) Option {
 }
 
+// WithContainerDiskAndPullPolicy specifies the disk name, the name of the container image and Pull Policy to be used.
+func WithContainerDiskAndPullPolicy(diskName, imageName string, imagePullPolicy k8sv1.PullPolicy) Option {
```
You could make `WithContainerDisk` call `WithContainerDiskAndPullPolicy` with an empty pull policy, so code duplication is avoided.

aliamerj replied:
call ??
do you mean change the name  ??

0xFelix commented on pkg/libvmi/storage.go:
```diff
@@ -118,6 +126,13 @@ func WithHostDisk(diskName, path string, diskType v1.HostDiskType) Option {
 	}
 }
 
+func WithHostDiskAndCapacity(diskName, path string, diskType v1.HostDiskType, capacity string) Option {
```
Same, make `WithHostDisk` call `WithHostDiskAndCapacity`? And pass `defaultDiskSize` to `WithHostDiskAndCapacity`.

0xFelix commented on pkg/libvmi/vmi.go:
```diff
@@ -263,6 +263,18 @@ func WithoutSerialConsole() Option {
 	}
 }
 
+func WithFirmware(imageName string) Option {
```
Should be called `WithKernelBootContainer`

0xFelix reviewed (commented):
A DCO signoff is missing in one of your commits. Also your commit messages look mangled.

aliamerj commented:
I don't know what just happened. I just want to help and complete this task this week. If I made any mistakes or wasted anyone's time, I am really sorry.

aliamerj commented:
I want to keep working on this project and I am planning to complete the task of refector all unit tests to use libvmi package this week 
So right now I am working on other files should I push it in the same pull request or open new one 
Make new commit or amend them 
Or waiting for the approval and merger the current commit and then push new one ??

fossedihelm commented:
> I want to keep working on this project and I am planning to complete the task of refector all unit tests to use libvmi package this week So right now I am working on other files should I push it in the same pull request or open new one Make new commit or amend them Or waiting for the approval and merger the current commit and then push new one ??

It is better to create separate PRs. This will help reviewers a lot. :) Thank you

aliamerj commented:
I accidentally closed this. I hope you still receive notifications about my pull request.
I fixed all issue with it

fossedihelm commented on pkg/libvmi/storage.go:
```diff
@@ -34,9 +34,14 @@ const (
+}
+
+// WithContainerDiskAndPullPolicy specifies the disk name, the name of the container image and Pull Policy to be used.
+func WithContainerDiskAndPullPolicy(diskName, imageName string, imagePullPolicy string) Option {
```
```suggestion
func WithContainerDiskAndPullPolicy(diskName, imageName string, imagePullPolicy k8sv1.PullPolicy) Option {
```
And save a cast at line 44

fossedihelm commented on pkg/libvmi/storage.go:
```diff
@@ -219,15 +228,20 @@ func newVolume(name string) v1.Volume {
 
-func newContainerVolume(name, image string) v1.Volume {
-	return v1.Volume{
+func newContainerVolume(name, image string, imagePullPolicy ...k8sv1.PullPolicy) v1.Volume {
```
```suggestion
func newContainerVolume(name, image string, imagePullPolicy k8sv1.PullPolicy) v1.Volume {
```
In this way you can directly check `imagePullPolicy != ""` later and prevent multiple imagePullPolices to be passed

fossedihelm commented on pkg/libvmi/storage.go:
```diff
@@ -219,15 +228,20 @@ func newVolume(name string) v1.Volume {
+		container.ContainerDisk.ImagePullPolicy = imagePullPolicy[0]
+	}
+	return container
+
```
drop empty new line

fossedihelm commented on pkg/libvmi/storage.go:
```diff
@@ -278,13 +292,15 @@ func newEmptyDisk(name string, capacity resource.Quantity) v1.Volume {
+	// Set capacity if provided, otherwise default to defaultDiskSize
 	if diskType == v1.HostDiskExistsOrCreate {
-		hostDisk.Capacity = resource.MustParse(defaultDiskSize)
+		hostDisk.Capacity = resource.MustParse(capacity)
```
This is wrong: you are setting the capacity only if `diskType == v1.HostDiskExistsOrCreate`.
I'd expect the capacity to be set if it is `!= ""`
Something like:
```
	// Set capacity if provided, otherwise default to defaultDiskSize
	if capacity != "" {
		hostDisk.Capacity = resource.MustParse(capacity)
	}
```

To have compatibility with both `WithHostDisk` and `WithHostDiskAndCapacity`, I'd expect to have something like:
```
func WithHostDisk(diskName, path string, diskType v1.HostDiskType) Option {
	var capacity string
	if diskType == v1.HostDiskExistsOrCreate {
		capacity = defaultDiskSize
	}
	return WithHostDiskAndCapacity(diskName, path, diskType, capacity)
}

func WithHostDiskAndCapacity(diskName, path string, diskType v1.HostDiskType, capacity string) Option {
	return func(vmi *v1.VirtualMachineInstance) {
		addDisk(vmi, newDisk(diskName, v1.DiskBusVirtio))
		addVolume(vmi, newHostDisk(diskName, path, diskType, capacity))
	}
}
```

aliamerj replied:
you are right

fossedihelm reviewed (commented):
Thank you: some comments below. PTAL

aliamerj commented:
@fossedihelm  
I have addressed all the feedback.
Thank you so much for consistently reviewing my code and providing valuable feedback. It has been incredibly helpful, and I truly appreciate it.

I hope I have not taken up too much of your time.

fossedihelm commented:
@aliamerj Did you change smth to the PR description template? There is the release note label required so I think you changed smth:
Maybe you dropped `release-note`?
```
"```release-note
NONE
```"
```

fossedihelm commented:
In any case I think release note is not required here: just put a `NONE` :)

lyarwood commented:
/area-remove instancetype

aliamerj commented:
Hello @fossedihelm 
is there anything i need to change in this PR ??

fossedihelm commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -182,8 +191,11 @@ var _ = Describe("ContainerDisk", func() {
-				vmi := api.NewMinimalVMI("fake-vmi")
-				appendContainerDisk(vmi, "r0")
+				vmi := libvmi.New(
+					libvmi.WithContainerDisk("r0", someImage),
```
you can use `WithResourceMemory`, `WithResourceCPU`, `WithLimitMemory`, `WithLimitCPU` and drop following resource and limits assignment

fossedihelm reviewed (commented):
One more comment from my side

aliamerj commented:
@fossedihelm 
I addressed it 
thank you for feedback : )

fossedihelm reviewed (approved):
Looks good to me! Thank you

fossedihelm commented:
/test all

aliamerj commented:
/test all

aliamerj commented:
/ok-to-test

aliamerj commented:
There was conflict in this PR and it was labeled as `needs-rebase`
So I resolved it using GitHub editor and this made 2 comments 
Is everything is okay ??

fossedihelm commented:
We don't use merge. You need to rebase your branch on main and force push

aliamerj commented:
@fossedihelm 
If I rebased the 2 commits into one 
Won't this make my commit mix with commit `Merge branch 'main'  into... '
Is this okay or do I need to remove the new commit ?

fossedihelm commented:
@aliamerj If you rebase I can provide my final feedback. Thank you

aliamerj commented:
@fossedihelm 
I believe i rebased it

fossedihelm commented on pkg/host-disk/host-disk_test.go:
```diff
@@ -379,32 +348,13 @@ var _ = Describe("HostDisk", func() {
+			vmi = libvmi.New(
+				libvmi.WithPersistentVolumeClaim(volumeName, pvcName),
+			)
+			vmi.Status.VolumeStatus = []v1.VolumeStatus{
```
we now have `WithVolumeStatus` :)

fossedihelm replied:
```suggestion
			vmi = libvmi.New(
				libvmi.WithPersistentVolumeClaim(volumeName, pvcName),
				libvmistatus.WithStatus(libvmistatus.New(
					libvmistatus.WithVolumeStatus(v1.VolumeStatus{
						Name:                      volumeName,
						PersistentVolumeClaimInfo: &v1.PersistentVolumeClaimInfo{},
					}),
				)),
			)
```

:)

aliamerj replied:
sorry i forgot this

fossedihelm reviewed (commented):
Thank you for your patience! Some comments below

aliamerj commented:
dropped it

fossedihelm reviewed (commented):
/test all
/lgtm
Thank you

alicefr commented:
@aliamerj would be possible to avoid all the additional prints? They are redundant and just make the change larget.
The rest looks good

fossedihelm commented on pkg/libvmi/vmi.go:
```diff
@@ -269,6 +271,24 @@ func WithoutSerialConsole() Option {
 	}
 }
 
+func WithActivePods(pods map[types.UID]string) Option {
+	return func(vmi *v1.VirtualMachineInstance) {
+		vmi.Status.ActivePods = pods
+	}
+}
```
We have `libvmistatus` now, so this is duplicated.
The libvmistatus can be used with:
`vmi := libvmi.New(libvmistatus.WithStatus(libvmistatus.New(libvmistatus.WithActivePod("uid", "nodename"))))

jean-edouard commented on pkg/container-disk/container-disk_test.go:
```diff
@@ -92,19 +92,25 @@ var _ = Describe("ContainerDisk", func() {
+				volumeMountDir := GetVolumeMountDirOnGuest(vmi)
+				_, err := os.Stat(filepath.Join(volumeMountDir, "disk_0.img"))
+				Expect(err).To(HaveOccurred())
+				Expect(err).To(MatchError(os.IsNotExist, "IsNotExist"))
```
The second Expect should be sufficient. If not error occurred, it won't match.
More curiously though, this test effectively tested nothing before?!

jean-edouard reviewed (commented):
Really nice cleanup, thank you! One tiny nit in-line, honestly not worth a new revision on its own.
/approve
